File: Sources/KiboCore/Models/ConversionMode.swift
import Foundation

/// How to interpret the text the user handed us.
///
/// Three of the four are mechanical: every mapped character is flipped, no questions asked.
/// `mixed` is the one that judges — see `KeyboardConverter` — so the other three double as the
/// escape hatch for when that judgement is wrong.
/// **Declaration order is the order of the picker**, since the control is built from `allCases`,
/// and it also sets the ⌘1–⌘4 shortcuts. It runs most-used first — a UI decision living in Core
/// only because `CaseIterable` puts it here.
public enum ConversionMode: String, CaseIterable, Sendable {
    /// Flip **every** run, each in the direction implied by the script it is already in — Thai
    /// runs back to English, Latin runs to Thai — without consulting the gate.
    ///
    /// This exists because text can be mistyped in *both* directions at once: switch layout
    /// halfway through a sentence and half of it is Thai-on-QWERTY while the other half is
    /// English-on-Kedmanee. Neither explicit direction can fix that, since each leaves the other
    /// script alone, and Mixed will not, because telling the two apart by spelling shape is
    /// measurably impossible — 36% of English words flip to well-formed Thai. Here the user
    /// supplies the judgement the gate cannot, so there is nothing left to guess wrong.
    case swapAll
    /// Treat the whole string as English keystrokes typed with the Thai layout active.
    case englishToThai
    /// Treat the whole string as Thai keystrokes typed with the US layout active.
    case thaiToEnglish
    /// Convert each run only if it is malformed in its own script; leave correct text alone.
    case mixed

    /// What the converter opens in when nothing is remembered — a fresh install, or a stored value
    /// that no longer parses.
    ///
    /// One constant rather than a literal at each of the three sites that need it, so the answer
    /// cannot be changed in two of them. It is deliberately **not** derived from `allCases.first`:
    /// the picker order is presentation and may be reshuffled again, while this is behaviour.
    ///
    /// `swapAll` rather than the safer `mixed`, by explicit product decision. It converts correct
    /// text, so a first-time user who pastes something already right will see it mangled — the
    /// trade is that the mode which always does *something* beats one that silently does nothing,
    /// and the result field is a preview the user reads before copying. Anyone who has used the
    /// app before is unaffected: `lastMode` is what they get.
    public static let `default`: ConversionMode = .swapAll

    /// The opposite explicit direction. `mixed` and `swapAll` are direction-symmetric and return
    /// themselves, so a Swap control can call this unconditionally.
    public var swapped: ConversionMode {
        switch self {
        case .mixed: return .mixed
        case .englishToThai: return .thaiToEnglish
        case .thaiToEnglish: return .englishToThai
        case .swapAll: return .swapAll
        }
    }

    /// Whether the mode has an opposite for a Swap control to flip to.
    public var hasDirection: Bool { swapped != self }
}

File: Windows/Kibo.Core/Models/ConversionMode.cs
namespace Kibo.Core;

/// <summary>
/// How to interpret the text the user handed us. The port of <c>ConversionMode.swift</c>.
/// </summary>
/// <remarks>
/// Three of the four are mechanical: every mapped character is flipped, no questions asked.
/// <see cref="Mixed"/> is the one that judges — see <see cref="KeyboardConverter"/> — so the other
/// three double as the escape hatch for when that judgement is wrong.
/// <para>
/// <b>Declaration order is the order of the picker</b>, via <see cref="ConversionModes.All"/>, and
/// it also sets the Ctrl+1–4 shortcuts. It runs most-used first.
/// </para>
/// </remarks>
public enum ConversionMode
{
    /// <summary>
    /// Flip <b>every</b> run, each in the direction implied by the script it is already in — Thai
    /// runs back to English, Latin runs to Thai — without consulting the gate. Exists because text
    /// can be mistyped in <i>both</i> directions at once, which no single explicit direction fixes
    /// and which Mixed provably cannot tell apart by spelling shape.
    /// </summary>
    SwapAll,

    /// <summary>Treat the whole string as English keystrokes typed with the Thai layout active.</summary>
    EnglishToThai,

    /// <summary>Treat the whole string as Thai keystrokes typed with the US layout active.</summary>
    ThaiToEnglish,

    /// <summary>Convert each run only if it is malformed in its own script; leave correct text alone.</summary>
    Mixed,
}

/// <summary>
/// What Swift hangs off the enum itself — <c>allCases</c>, <c>rawValue</c>, <c>default</c>,
/// <c>swapped</c> — lives here, because a C# enum cannot carry members.
/// </summary>
public static class ConversionModes
{
    /// <summary>Picker order. Most-used first.</summary>
    public static IReadOnlyList<ConversionMode> All { get; } =
        [ConversionMode.SwapAll, ConversionMode.EnglishToThai, ConversionMode.ThaiToEnglish, ConversionMode.Mixed];

    /// <summary>
    /// What the converter opens in when nothing is remembered — a fresh install, or a stored value
    /// that no longer parses. Deliberately <b>not</b> <c>All[0]</c>: picker order is presentation
    /// and may be reshuffled, while this is behaviour. <c>SwapAll</c> rather than the safer
    /// <c>Mixed</c> by explicit product decision — the result field is a preview the user reads
    /// before copying, and a mode that always does <i>something</i> beats one that silently does
    /// nothing.
    /// </summary>
    public const ConversionMode Default = ConversionMode.SwapAll;

    /// <summary>
    /// The opposite explicit direction. <c>Mixed</c> and <c>SwapAll</c> are direction-symmetric and
    /// return themselves, so a Swap control can call this unconditionally.
    /// </summary>
    public static ConversionMode Swapped(this ConversionMode mode) => mode switch
    {
        ConversionMode.EnglishToThai => ConversionMode.ThaiToEnglish,
        ConversionMode.ThaiToEnglish => ConversionMode.EnglishToThai,
        _ => mode,
    };

    /// <summary>Whether the mode has an opposite for a Swap control to flip to.</summary>
    public static bool HasDirection(this ConversionMode mode) => mode.Swapped() != mode;

    /// <summary>
    /// The identifier used by <c>Fixtures/conversion-cases.json</c> and by the settings file. Fixed
    /// by that contract, not by the C# member names.
    /// </summary>
    public static string RawValue(this ConversionMode mode) => mode switch
    {
        ConversionMode.SwapAll => "swapAll",
        ConversionMode.EnglishToThai => "englishToThai",
        ConversionMode.ThaiToEnglish => "thaiToEnglish",
        ConversionMode.Mixed => "mixed",
        _ => throw new ArgumentOutOfRangeException(nameof(mode), mode, null),
    };

    /// <summary>
    /// Exact, ordinal parsing of a raw value. Hand-written rather than <see cref="Enum.TryParse{TEnum}(string?, out TEnum)"/>,
    /// which accepts integers ("3") and, with the wrong flag, any casing — a corrupted or
    /// newer-build setting must fall back to <see cref="Default"/>, not land on a mode by accident.
    /// </summary>
    public static bool TryParse(string? raw, out ConversionMode mode)
    {
        foreach (var candidate in All)
        {
            if (string.Equals(candidate.RawValue(), raw, StringComparison.Ordinal))
            {
                mode = candidate;
                return true;
            }
        }
        mode = Default;
        return false;
    }
}
